File: src/main/java/top/ceroxe/api/neolink/network/threads/UDPTransformer.java
package top.ceroxe.api.neolink.network.threads;

import top.ceroxe.api.net.SecureSocket;
import top.ceroxe.api.neolink.util.Debugger;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;
import java.util.Objects;
import java.util.function.BiConsumer;

import static top.ceroxe.api.neolink.network.InternetOperator.close;

/**
 * UDP 数据转发器。
 *
 * 核心职责：
 * 1. 在本地服务和 Neo 服务器之间双向转发 UDP 数据
 * 2. 使用 ByteBuffer 实现高效的字节操作
 * 3. 通过复用实例缓冲区减少 GC 压力
 *
 * 设计特点：
 * - 双向转发：支持 Neo -> 本地、本地 -> Neo 两种模式
 * - ByteBuffer 优化：使用堆上缓冲区提升 I/O 性能
 * - 缓冲区复用：每个实例使用独立缓冲区，避免频繁分配内存
 * - 优雅关闭：支持中断信号，确保资源正确释放
 *
 * 性能优化：
 * - 使用 65535 字节缓冲区，支持最大 UDP 报文
 * - 通过 ByteBuffer 直接操作字节数组，减少拷贝开销
 *
 * @author NeoProxy Team
 * @since 5.0.0
 */
public class UDPTransformer implements Runnable {
    public static final int MODE_NEO_TO_LOCAL = 0;
    public static final int MODE_LOCAL_TO_NEO = 1;
    private static final int MAGIC = 0xDEADBEEF;
    private static final int SERIALIZED_HEADER_FIXED_LENGTH = 4 + 4 + 4 + 2;
    private static final int IPV4_LENGTH = 4;
    private static final int IPV6_LENGTH = 16;
    private static final int BUFFER_LENGTH = 65535;

    private final DatagramSocket plainSocket;
    private final SecureSocket secureSocket;
    private final int mode;
    private final String localHost;
    private final InetAddress localAddress;
    private final int localPort;
    private final boolean debugEnabled;
    private final BiConsumer<String, Throwable> debugSink;

    // 每条 UDP 转发链路独占接收缓冲区，避免多连接并发时复用同一数组。
    private final byte[] receiveBuffer = new byte[BUFFER_LENGTH];

    // 序列化缓冲区绑定在实例上，既减少分配，也避免不同 UDP 会话互相覆盖数据。
    private final ByteBuffer serializationBuffer = ByteBuffer.allocate(
            SERIALIZED_HEADER_FIXED_LENGTH + IPV6_LENGTH + BUFFER_LENGTH
    );

    /**
     * 构造函数：用于从 Neo 服务器接收数据并转发到本地服务。
     */
    public UDPTransformer(SecureSocket secureSender, DatagramSocket localReceiver) {
        this(secureSender, localReceiver, "localhost", 1);
    }

    public UDPTransformer(SecureSocket secureSender, DatagramSocket localReceiver, String localHost, int localPort) {
        this(secureSender, localReceiver, localHost, localPort, Debugger.isEnabled());
    }

    public UDPTransformer(
            SecureSocket secureSender,
            DatagramSocket localReceiver,
            String localHost,
            int localPort,
            boolean debugEnabled
    ) {
        this(secureSender, localReceiver, localHost, localPort, debugEnabled, UDPTransformer::defaultDebugSink);
    }

    public UDPTransformer(
            SecureSocket secureSender,
            DatagramSocket localReceiver,
            String localHost,
            int localPort,
            boolean debugEnabled,
            BiConsumer<String, Throwable> debugSink
    ) {
        this.secureSocket = secureSender;
        this.plainSocket = localReceiver;
        this.mode = MODE_NEO_TO_LOCAL;
        this.localHost = localHost;
        this.localAddress = resolveLocalAddress(localHost);
        this.localPort = localPort;
        this.debugEnabled = debugEnabled;
        this.debugSink = Objects.requireNonNull(debugSink, "debugSink");
    }

    /**
     * 构造函数：用于从本地服务接收数据并转发到 Neo 服务器。
     */
    public UDPTransformer(DatagramSocket localSender, SecureSocket secureReceiver) {
        this(localSender, secureReceiver, "localhost", 1);
    }

    public UDPTransformer(DatagramSocket localSender, SecureSocket secureReceiver, String localHost, int localPort) {
        this(localSender, secureReceiver, localHost, localPort, Debugger.isEnabled());
    }

    public UDPTransformer(
            DatagramSocket localSender,
            SecureSocket secureReceiver,
            String localHost,
            int localPort,
            boolean debugEnabled
    ) {
        this(localSender, secureReceiver, localHost, localPort, debugEnabled, UDPTransformer::defaultDebugSink);
    }

    public UDPTransformer(
            DatagramSocket localSender,
            SecureSocket secureReceiver,
            String localHost,
            int localPort,
            boolean debugEnabled,
            BiConsumer<String, Throwable> debugSink
    ) {
        this.plainSocket = localSender;
        this.secureSocket = secureReceiver;
        this.mode = MODE_LOCAL_TO_NEO;
        this.localHost = localHost;
        this.localAddress = resolveLocalAddress(localHost);
        this.localPort = localPort;
        this.debugEnabled = debugEnabled;
        this.debugSink = Objects.requireNonNull(debugSink, "debugSink");
    }

    /**
     * 这个方法可以保持为静态，因为它不依赖实例状态。
     */
    public static DatagramPacket deserializeToDatagramPacket(byte[] serializedData) {
        return deserializeToDatagramPacket(serializedData, true, UDPTransformer::defaultDebugSink);
    }

    static DatagramPacket deserializeToDatagramPacket(
            byte[] serializedData,
            boolean debugEnabled,
            BiConsumer<String, Throwable> debugSink
    ) {
        if (serializedData == null || serializedData.length < SERIALIZED_HEADER_FIXED_LENGTH + IPV4_LENGTH) {
            return invalidSerializedPacket("Serialized UDP packet is too short", debugEnabled, debugSink);
        }

        ByteBuffer buffer = ByteBuffer.wrap(serializedData);
        buffer.order(ByteOrder.BIG_ENDIAN);

        int magic = buffer.getInt();
        if (magic != MAGIC) {
            return invalidSerializedPacket("Invalid magic number in serialized data", debugEnabled, debugSink);
        }

        int dataLen = buffer.getInt();
        int ipLen = buffer.getInt();
        if (dataLen < 0 || dataLen > BUFFER_LENGTH) {
            return invalidSerializedPacket("Invalid UDP data length: " + dataLen, debugEnabled, debugSink);
        }
        if (ipLen != IPV4_LENGTH && ipLen != IPV6_LENGTH) {
            return invalidSerializedPacket("Invalid IP address length: " + ipLen, debugEnabled, debugSink);
        }

        int expectedLength = SERIALIZED_HEADER_FIXED_LENGTH + ipLen + dataLen;
        if (serializedData.length != expectedLength) {
            return invalidSerializedPacket("Serialized UDP packet length mismatch", debugEnabled, debugSink);
        }

        byte[] ipBytes = new byte[ipLen];
        buffer.get(ipBytes);
        InetAddress address;
        try {
            address = InetAddress.getByAddress(ipBytes);
        } catch (Exception e) {
            emitMalformedPacketDebug(null, e, debugEnabled, debugSink);
            return null;
        }
        int port = buffer.getShort() & 0xFFFF;
        byte[] data = new byte[dataLen];
        buffer.get(data);

        return new DatagramPacket(data, data.length, address, port);
    }

    private static DatagramPacket invalidSerializedPacket(
            String message,
            boolean debugEnabled,
            BiConsumer<String, Throwable> debugSink
    ) {
        emitMalformedPacketDebug(message, null, debugEnabled, debugSink);
        return null;
    }

    private void transferDataToNeoServer() {
        try {
            while (true) { // 用异常退出循环。
                DatagramPacket incomingPacket = new DatagramPacket(receiveBuffer, receiveBuffer.length);
                plainSocket.receive(incomingPacket);
                byte[] serializedData = serializeDatagramPacket(incomingPacket);
                debug("Forwarding UDP packet from local service to NeoProxyServer. source="
                        + incomingPacket.getAddress().getHostAddress() + ":" + incomingPacket.getPort()
                        + ", payloadBytes=" + incomingPacket.getLength()
                        + ", serializedBytes=" + serializedData.length);
                secureSocket.sendBytes(serializedData);
            }
        } catch (IOException e) {
            debug(e);
        }
    }

    private byte[] serializeDatagramPacket(DatagramPacket packet) {
        serializationBuffer.clear();
        serializationBuffer.order(ByteOrder.BIG_ENDIAN);

        byte[] data = packet.getData();
        int offset = packet.getOffset();
        int length = packet.getLength();
        InetAddress address = packet.getAddress();
        int port = packet.getPort();

        byte[] ipBytes = address.getAddress();
        int ipLength = ipBytes.length;

        // 先校验序列化缓冲区容量是否足够，避免静默截断大包。
        // 这里选择直接抛异常，让调用方明确知道包太大。
        int totalLen = 4 + 4 + 4 + ipLength + 2 + length;
        if (totalLen > serializationBuffer.capacity()) {
            throw new IllegalArgumentException("UDP packet too large for serialization buffer: " + totalLen);
        }

        serializationBuffer.putInt(MAGIC);
        serializationBuffer.putInt(length);
        serializationBuffer.putInt(ipLength);
        serializationBuffer.put(ipBytes);
        serializationBuffer.putShort((short) port);
        serializationBuffer.put(data, offset, length);

        // 返回副本，避免后续复用 serializationBuffer 时覆盖正在发送的数据。
        return Arrays.copyOf(serializationBuffer.array(), serializationBuffer.position());
    }

    private void transferDataToLocalServer() {
        try {
            byte[] data;
            while ((data = secureSocket.receiveBytes()) != null) {
                debug("Received UDP frame from NeoProxyServer. serializedBytes=" + data.length);
                DatagramPacket datagramPacket = deserializeToDatagramPacket(data, debugEnabled, this::emitDebug);
                if (datagramPacket != null) {
                    DatagramPacket outgoingPacket = new DatagramPacket(
                            datagramPacket.getData(),
                            datagramPacket.getLength(),
                            localAddress,
                            localPort
                    );
                    plainSocket.send(outgoingPacket);
                    debug("Forwarded UDP packet to local service. target=" + localHost + ":" + localPort
                            + ", payloadBytes=" + datagramPacket.getLength());
                }
            }
        } catch (Exception e) {
            debug(e);
        }
    }

    @Override
    public void run() {
        try {
            debug("UDP transformer started. mode=" + (mode == MODE_NEO_TO_LOCAL ? "NEO_TO_LOCAL" : "LOCAL_TO_NEO")
                    + ", localTarget=" + localHost + ":" + localPort);
            if (mode == MODE_NEO_TO_LOCAL) {
                transferDataToLocalServer();
            } else {
                transferDataToNeoServer();
            }
        } catch (Exception e) {
            debug(e);
        } finally {
            // 无论正常结束还是异常结束，都确保走到统一的资源收尾路径。
            debug("UDP transformer finished.");
        }
    }

    private void debug(String message) {
        if (!debugEnabled) {
            return;
        }
        emitDebug(message, null);
    }

    private void debug(Exception e) {
        if (!debugEnabled || e == null) {
            return;
        }
        emitDebug(null, e);
    }

    private void emitDebug(String message, Throwable cause) {
        try {
            debugSink.accept(message, cause);
        } catch (RuntimeException ignored) {
            // Debug 回调只用于观测，不能干扰转发流程。
        }
    }

    private static void emitMalformedPacketDebug(
            String message,
            Throwable cause,
            boolean debugEnabled,
            BiConsumer<String, Throwable> debugSink
    ) {
        if (!debugEnabled) {
            return;
        }
        try {
            debugSink.accept(message, cause);
        } catch (RuntimeException ignored) {
            // 畸形包诊断只用于观测，不能干扰转发流程。
        }
    }

    private static InetAddress resolveLocalAddress(String localHost) {
        try {
            return InetAddress.getByName(localHost);
        } catch (IOException e) {
            throw new UncheckedIOException("Unable to resolve local UDP target host: " + localHost, e);
        }
    }

    private static void defaultDebugSink(String message, Throwable cause) {
        if (message != null) {
            Debugger.debugOperation(true, message);
        }
        if (cause instanceof Exception exception) {
            Debugger.debugOperation(true, exception);
        }
    }
}

File: packages/nodejs/src/network/udp-transformer.ts
import dgram from 'node:dgram';
import net from 'node:net';
import {SecureSocket} from '../secure-socket.js';
import type {DebugSink} from '../types.js';

export const MAGIC = 0xDEADBEEF;
export const SERIALIZED_HEADER_FIXED_LENGTH = 4 + 4 + 4 + 2;
export const IPV4_LENGTH = 4;
export const IPV6_LENGTH = 16;
export const BUFFER_LENGTH = 65535;

export interface DatagramPacket {
    data: Buffer;
    address: string;
    port: number;
}

export function ipv4ToBytes(address: string): Buffer {
    const parts = address.split('.').map((part) => Number(part));
    if (parts.length !== 4 || parts.some((part) => !Number.isInteger(part) || part < 0 || part > 255)) {
        throw new Error(`Invalid IPv4 address: ${address}`);
    }
    return Buffer.from(parts);
}

export function ipv6ToBytes(address: string): Buffer {
    let value = address;
    const zoneIndex = value.indexOf('%');
    if (zoneIndex >= 0) {
        value = value.slice(0, zoneIndex);
    }
    if (value.includes('.')) {
        const lastColon = value.lastIndexOf(':');
        const ipv4 = ipv4ToBytes(value.slice(lastColon + 1));
        value = `${value.slice(0, lastColon)}:${ipv4.readUInt16BE(0).toString(16)}:${ipv4.readUInt16BE(2).toString(16)}`;
    }
    const hasCompression = value.includes('::');
    const [headRaw, tailRaw = ''] = value.split('::');
    const head = headRaw ? headRaw.split(':').filter(Boolean) : [];
    const tail = tailRaw ? tailRaw.split(':').filter(Boolean) : [];
    const missing = hasCompression ? 8 - head.length - tail.length : 0;
    const parts = hasCompression ? [...head, ...Array(missing).fill('0'), ...tail] : head;
    if (parts.length !== 8 || missing < 0) {
        throw new Error(`Invalid IPv6 address: ${address}`);
    }
    const out = Buffer.alloc(16);
    parts.forEach((part, index) => {
        const value16 = parseInt(part, 16);
        if (!Number.isInteger(value16) || value16 < 0 || value16 > 0xFFFF) {
            throw new Error(`Invalid IPv6 address: ${address}`);
        }
        out.writeUInt16BE(value16, index * 2);
    });
    return out;
}

export function ipToBytes(address: string): Buffer {
    const family = net.isIP(address);
    if (family === 4) {
        return ipv4ToBytes(address);
    }
    if (family === 6) {
        return ipv6ToBytes(address);
    }
    throw new Error(`Invalid IP address: ${address}`);
}

export function bytesToIp(bytes: Buffer): string {
    if (bytes.length === IPV4_LENGTH) {
        return Array.from(bytes).join('.');
    }
    if (bytes.length === IPV6_LENGTH) {
        const parts: string[] = [];
        for (let offset = 0; offset < bytes.length; offset += 2) {
            parts.push(bytes.readUInt16BE(offset).toString(16));
        }
        return parts.join(':');
    }
    throw new Error(`Invalid IP address length: ${bytes.length}`);
}

export function serializeDatagramPacket(message: Buffer | Uint8Array, rinfo: {
    address: string;
    port: number
}): Buffer {
    const payload = Buffer.from(message);
    const ipBytes = ipToBytes(rinfo.address);
    const totalLength = 4 + 4 + 4 + ipBytes.length + 2 + payload.length;
    if (totalLength > BUFFER_LENGTH) {
        throw new RangeError(`UDP packet too large for serialization buffer: ${totalLength}`);
    }
    const out = Buffer.alloc(totalLength);
    let offset = 0;
    out.writeUInt32BE(MAGIC, offset);
    offset += 4;
    out.writeInt32BE(payload.length, offset);
    offset += 4;
    out.writeInt32BE(ipBytes.length, offset);
    offset += 4;
    ipBytes.copy(out, offset);
    offset += ipBytes.length;
    out.writeUInt16BE(rinfo.port, offset);
    offset += 2;
    payload.copy(out, offset);
    return out;
}

export function deserializeToDatagramPacket(serializedData: Buffer | Uint8Array, debug: DebugSink = () => undefined): DatagramPacket | null {
    const data = Buffer.from(serializedData);
    if (data.length < SERIALIZED_HEADER_FIXED_LENGTH + IPV4_LENGTH) {
        debug('Serialized UDP packet is too short');
        return null;
    }
    let offset = 0;
    const magic = data.readUInt32BE(offset);
    offset += 4;
    if (magic !== MAGIC) {
        debug('Invalid magic number in serialized data');
        return null;
    }
    const dataLen = data.readInt32BE(offset);
    offset += 4;
    const ipLen = data.readInt32BE(offset);
    offset += 4;
    if (dataLen < 0 || dataLen > BUFFER_LENGTH) {
        debug(`Invalid UDP data length: ${dataLen}`);
        return null;
    }
    if (ipLen !== IPV4_LENGTH && ipLen !== IPV6_LENGTH) {
        debug(`Invalid IP address length: ${ipLen}`);
        return null;
    }
    const expectedLength = SERIALIZED_HEADER_FIXED_LENGTH + ipLen + dataLen;
    if (data.length !== expectedLength) {
        debug('Serialized UDP packet length mismatch');
        return null;
    }
    const ipBytes = data.subarray(offset, offset + ipLen);
    offset += ipLen;
    const port = data.readUInt16BE(offset);
    offset += 2;
    const payload = Buffer.from(data.subarray(offset, offset + dataLen));
    return {data: payload, address: bytesToIp(ipBytes), port};
}

interface DatagramQueue {
    next(): Promise<{ message: Buffer; rinfo: dgram.RemoteInfo } | null>;
}

function createDatagramQueue(socket: dgram.Socket): DatagramQueue {
    const queue: Array<{ message: Buffer; rinfo: dgram.RemoteInfo }> = [];
    const waiters: Array<{
        resolve: (value: { message: Buffer; rinfo: dgram.RemoteInfo } | null) => void;
        reject: (error: Error) => void
    }> = [];
    let closed = false;
    let failure: Error | null = null;
    socket.on('message', (message, rinfo) => {
        const item = {message, rinfo};
        const waiter = waiters.shift();
        if (waiter) {
            waiter.resolve(item);
        } else {
            queue.push(item);
        }
    });
    socket.once('close', () => {
        closed = true;
        for (const waiter of waiters.splice(0)) {
            waiter.resolve(null);
        }
    });
    socket.once('error', (error) => {
        failure = error;
        for (const waiter of waiters.splice(0)) {
            waiter.reject(error);
        }
    });
    return {
        next() {
            if (queue.length > 0) {
                return Promise.resolve(queue.shift()!);
            }
            if (failure) {
                return Promise.reject(failure);
            }
            if (closed) {
                return Promise.resolve(null);
            }
            return new Promise((resolve, reject) => waiters.push({resolve, reject}));
        }
    };
}

export function sendDatagram(socket: dgram.Socket, message: Buffer | Uint8Array, port: number, host: string): Promise<void> {
    return new Promise((resolve, reject) => {
        socket.send(message, port, host, (error) => error ? reject(error) : resolve());
    });
}

export async function transferLocalToNeo(datagramSocket: dgram.Socket, secureSocket: SecureSocket, debug: DebugSink = () => undefined): Promise<void> {
    const queue = createDatagramQueue(datagramSocket);
    try {
        while (true) {
            const item = await queue.next();
            if (item == null) {
                break;
            }
            const serialized = serializeDatagramPacket(item.message, item.rinfo);
            debug(`Forwarding UDP packet from local service to NeoProxyServer. source=${item.rinfo.address}:${item.rinfo.port}, payloadBytes=${item.message.length}, serializedBytes=${serialized.length}`);
            await secureSocket.sendBytes(serialized);
        }
    } catch (error) {
        debug(null, error);
    }
}

export async function transferNeoToLocal(
    secureSocket: SecureSocket,
    datagramSocket: dgram.Socket,
    localHost: string,
    localPort: number,
    debug: DebugSink = () => undefined
): Promise<void> {
    try {
        while (true) {
            const data = await secureSocket.receiveBytes();
            if (data == null) {
                break;
            }
            debug(`Received UDP frame from NeoProxyServer. serializedBytes=${data.length}`);
            const packet = deserializeToDatagramPacket(data, debug);
            if (packet) {
                await sendDatagram(datagramSocket, packet.data, localPort, localHost);
                debug(`Forwarded UDP packet to local service. target=${localHost}:${localPort}, payloadBytes=${packet.data.length}`);
            }
        }
    } catch (error) {
        debug(null, error);
    }
}

export function createUdpSocket(): Promise<dgram.Socket> {
    const socket = dgram.createSocket('udp4');
    return new Promise((resolve, reject) => {
        socket.once('error', reject);
        socket.bind(0, '0.0.0.0', () => {
            socket.off('error', reject);
            resolve(socket);
        });
    });
}
